File: site/app/detection/geometry.ts
import type { Point, Quad } from "./types.ts";
import { DETECTION_CONFIG } from "./config.ts";

const GEOMETRY_CONFIG = DETECTION_CONFIG.geometry;

export type Line = {
  a: number;
  b: number;
  c: number;
};

export function lineIntersection(first: Line, second: Line): Point {
  const denominator = first.a * second.b - second.a * first.b;
  if (Math.abs(denominator) < 1e-9) return [Number.NaN, Number.NaN];
  return [
    (first.b * second.c - second.b * first.c) / denominator,
    (first.c * second.a - second.c * first.a) / denominator,
  ];
}

export function orderQuad(points: Point[]): Quad {
  const bySum = [...points].sort((a, b) => a[0] + a[1] - (b[0] + b[1]));
  const byDiff = [...points].sort((a, b) => a[0] - a[1] - (b[0] - b[1]));
  return [bySum[0], byDiff[3], bySum[3], byDiff[0]] as Quad;
}

export function polygonArea(points: Quad): number {
  let area = 0;
  for (let index = 0; index < points.length; index += 1) {
    const current = points[index];
    const next = points[(index + 1) % points.length];
    area += current[0] * next[1] - current[1] * next[0];
  }
  return Math.abs(area) * 0.5;
}

export function distance(first: Point, second: Point): number {
  return Math.hypot(first[0] - second[0], first[1] - second[1]);
}

export function isPointInside(point: Point, quad: Quad): boolean {
  let inside = false;
  for (let index = 0, previous = quad.length - 1; index < quad.length; previous = index, index += 1) {
    const [x1, y1] = quad[index];
    const [x2, y2] = quad[previous];
    if ((y1 > point[1]) !== (y2 > point[1])) {
      const crossingX = ((x2 - x1) * (point[1] - y1)) / (y2 - y1) + x1;
      if (point[0] < crossingX) inside = !inside;
    }
  }
  return inside;
}

export function isConvexQuad(quad: Quad): boolean {
  const signs: number[] = [];
  for (let index = 0; index < 4; index += 1) {
    const first = quad[index];
    const second = quad[(index + 1) % 4];
    const third = quad[(index + 2) % 4];
    const cross = (second[0] - first[0]) * (third[1] - second[1]) -
      (second[1] - first[1]) * (third[0] - second[0]);
    if (Math.abs(cross) < 1e-6) return false;
    signs.push(Math.sign(cross));
  }
  return signs.every((sign) => sign === signs[0]);
}

export function geometryIsValid(quad: Quad, width: number, height: number): boolean {
  if (!quad.every(([x, y]) => Number.isFinite(x) && Number.isFinite(y))) return false;
  if (!isConvexQuad(quad)) return false;
  if (polygonArea(quad) < width * height * GEOMETRY_CONFIG.minimumAreaRatio) return false;
  if (quad.some(([x, y]) =>
    x < -width * GEOMETRY_CONFIG.boundsRatio ||
    x > width * (1 + GEOMETRY_CONFIG.boundsRatio) ||
    y < -height * GEOMETRY_CONFIG.boundsRatio ||
    y > height * (1 + GEOMETRY_CONFIG.boundsRatio)
  )) {
    return false;
  }
  const shortestEdge = Math.min(...quad.map((point, index) => distance(point, quad[(index + 1) % 4])));
  if (shortestEdge < Math.min(width, height) * GEOMETRY_CONFIG.minimumEdgeRatio) return false;
  for (let index = 0; index < 4; index += 1) {
    const previous = quad[(index + 3) % 4];
    const current = quad[index];
    const next = quad[(index + 1) % 4];
    const first = [previous[0] - current[0], previous[1] - current[1]];
    const second = [next[0] - current[0], next[1] - current[1]];
    const cosine = (first[0] * second[0] + first[1] * second[1]) /
      Math.max(1e-9, Math.hypot(...first) * Math.hypot(...second));
    const angle = Math.acos(Math.max(-1, Math.min(1, cosine))) * 180 / Math.PI;
    if (angle < GEOMETRY_CONFIG.minimumAngleDegrees || angle > GEOMETRY_CONFIG.maximumAngleDegrees) return false;
  }
  return true;
}

export function normalizedCornerDistance(first: Quad, second: Quad, width: number, height: number): number {
  const diagonal = Math.hypot(width, height);
  return first.reduce((sum, point, index) => sum + distance(point, second[index]), 0) / (4 * diagonal);
}

export function quadIoU(first: Quad, second: Quad, width: number, height: number): number {
  const scale = Math.max(1, Math.ceil(Math.max(width, height) / 240));
  let intersection = 0;
  let union = 0;
  for (let y = 0; y < height; y += scale) {
    for (let x = 0; x < width; x += scale) {
      const point: Point = [x + scale / 2, y + scale / 2];
      const insideFirst = isPointInside(point, first);
      const insideSecond = isPointInside(point, second);
      if (insideFirst || insideSecond) union += 1;
      if (insideFirst && insideSecond) intersection += 1;
    }
  }
  return union ? intersection / union : 0;
}

export function convexQuadIoU(first: Quad, second: Quad): number {
  const intersection = clipConvexPolygon(first, second);
  const intersectionArea = polygonAreaPoints(intersection);
  const unionArea = polygonArea(first) + polygonArea(second) - intersectionArea;
  return unionArea > 0 ? intersectionArea / unionArea : 0;
}

export function scaleQuad(quad: Quad, factor: number): Quad {
  const center: Point = [
    quad.reduce((sum, point) => sum + point[0], 0) / 4,
    quad.reduce((sum, point) => sum + point[1], 0) / 4,
  ];
  return quad.map(([x, y]) => [
    center[0] + (x - center[0]) * factor,
    center[1] + (y - center[1]) * factor,
  ]) as Quad;
}

function clipConvexPolygon(subject: Point[], clip: Point[]): Point[] {
  let output = [...subject];
  const orientation = signedPolygonArea(clip) >= 0 ? 1 : -1;
  for (let edgeIndex = 0; edgeIndex < clip.length; edgeIndex += 1) {
    const edgeStart = clip[edgeIndex];
    const edgeEnd = clip[(edgeIndex + 1) % clip.length];
    const input = output;
    output = [];
    if (!input.length) break;
    let previous = input[input.length - 1];
    let previousInside = halfPlane(previous, edgeStart, edgeEnd) * orientation >= -1e-7;
    for (const current of input) {
      const currentInside = halfPlane(current, edgeStart, edgeEnd) * orientation >= -1e-7;
      if (currentInside) {
        if (!previousInside) output.push(segmentLineIntersection(previous, current, edgeStart, edgeEnd));
        output.push(current);
      } else if (previousInside) {
        output.push(segmentLineIntersection(previous, current, edgeStart, edgeEnd));
      }
      previous = current;
      previousInside = currentInside;
    }
  }
  return output;
}

function segmentLineIntersection(start: Point, end: Point, lineStart: Point, lineEnd: Point): Point {
  const segmentX = end[0] - start[0];
  const segmentY = end[1] - start[1];
  const lineX = lineEnd[0] - lineStart[0];
  const lineY = lineEnd[1] - lineStart[1];
  const denominator = segmentX * lineY - segmentY * lineX;
  if (Math.abs(denominator) < 1e-9) return end;
  const offsetX = lineStart[0] - start[0];
  const offsetY = lineStart[1] - start[1];
  const fraction = (offsetX * lineY - offsetY * lineX) / denominator;
  return [start[0] + segmentX * fraction, start[1] + segmentY * fraction];
}

function halfPlane(point: Point, start: Point, end: Point): number {
  return (end[0] - start[0]) * (point[1] - start[1]) -
    (end[1] - start[1]) * (point[0] - start[0]);
}

function polygonAreaPoints(points: Point[]): number {
  return Math.abs(signedPolygonArea(points));
}

function signedPolygonArea(points: Point[]): number {
  let area = 0;
  for (let index = 0; index < points.length; index += 1) {
    const current = points[index];
    const next = points[(index + 1) % points.length];
    area += current[0] * next[1] - current[1] * next[0];
  }
  return area * 0.5;
}

File: src/slides_thief/detection/geometry.py
"""Canonical quadrilateral geometry primitives for the detector."""

from __future__ import annotations

import math

import numpy as np

from .config import DETECTION_CONFIG

_GEOMETRY_CONFIG = DETECTION_CONFIG["geometry"]


def polygon_area(points: np.ndarray) -> float:
    """Return the absolute area of an ordered polygon."""

    if len(points) < 3:
        return 0.0
    return 0.5 * abs(
        float(np.dot(points[:, 0], np.roll(points[:, 1], -1))
        - np.dot(points[:, 1], np.roll(points[:, 0], -1)))
    )


def geometry_is_valid(quad: np.ndarray, width: int, height: int) -> bool:
    """Validate an ordered top-left, top-right, bottom-right, bottom-left quad."""

    if quad.shape != (4, 2) or not np.isfinite(quad).all():
        return False
    if polygon_area(quad) < width * height * float(_GEOMETRY_CONFIG["minimumAreaRatio"]):
        return False
    bounds_ratio = float(_GEOMETRY_CONFIG["boundsRatio"])
    if np.any(quad[:, 0] < -width * bounds_ratio) or np.any(quad[:, 0] > width * (1 + bounds_ratio)):
        return False
    if np.any(quad[:, 1] < -height * bounds_ratio) or np.any(quad[:, 1] > height * (1 + bounds_ratio)):
        return False

    edge_vectors = np.roll(quad, -1, axis=0) - quad
    edge_lengths = np.linalg.norm(edge_vectors, axis=1)
    if np.any(edge_lengths < min(width, height) * float(_GEOMETRY_CONFIG["minimumEdgeRatio"])):
        return False
    crosses = edge_vectors[:, 0] * np.roll(edge_vectors[:, 1], -1) - edge_vectors[:, 1] * np.roll(
        edge_vectors[:, 0], -1
    )
    if not (np.all(crosses > 1e-6) or np.all(crosses < -1e-6)):
        return False

    for index in range(4):
        previous = quad[(index + 3) % 4] - quad[index]
        following = quad[(index + 1) % 4] - quad[index]
        cosine = float(np.dot(previous, following) / max(1e-9, np.linalg.norm(previous) * np.linalg.norm(following)))
        angle = math.degrees(math.acos(max(-1.0, min(1.0, cosine))))
        if angle < float(_GEOMETRY_CONFIG["minimumAngleDegrees"]) or angle > float(
            _GEOMETRY_CONFIG["maximumAngleDegrees"]
        ):
            return False
    return True


def line_intersection(first: np.ndarray, second: np.ndarray) -> np.ndarray:
    """Intersect two homogeneous lines, returning NaN coordinates when parallel."""

    denominator = first[0] * second[1] - second[0] * first[1]
    if abs(denominator) < 1e-9:
        return np.array([np.nan, np.nan], dtype=np.float64)
    return np.array(
        [
            (first[1] * second[2] - second[1] * first[2]) / denominator,
            (first[2] * second[0] - second[2] * first[0]) / denominator,
        ],
        dtype=np.float64,
    )


def normalized_corner_distance(first: np.ndarray, second: np.ndarray, width: int, height: int) -> float:
    return float(np.linalg.norm(first - second, axis=1).mean() / math.hypot(width, height))


def quad_iou(first: np.ndarray, second: np.ndarray) -> float:
    """Return exact IoU for two convex quads using polygon clipping."""

    intersection = _clip_convex_polygon(first, second)
    intersection_area = polygon_area(intersection)
    union_area = polygon_area(first) + polygon_area(second) - intersection_area
    return intersection_area / union_area if union_area > 0 else 0.0


def _clip_convex_polygon(subject: np.ndarray, clip: np.ndarray) -> np.ndarray:
    output = [np.asarray(point, dtype=np.float64) for point in subject]
    orientation = 1.0 if _signed_polygon_area(clip) >= 0 else -1.0
    for edge_index in range(len(clip)):
        edge_start = clip[edge_index]
        edge_end = clip[(edge_index + 1) % len(clip)]
        input_points = output
        output = []
        if not input_points:
            break
        previous = input_points[-1]
        previous_inside = _half_plane(previous, edge_start, edge_end) * orientation >= -1e-7
        for current in input_points:
            current_inside = _half_plane(current, edge_start, edge_end) * orientation >= -1e-7
            if current_inside:
                if not previous_inside:
                    output.append(_segment_line_intersection(previous, current, edge_start, edge_end))
                output.append(current)
            elif previous_inside:
                output.append(_segment_line_intersection(previous, current, edge_start, edge_end))
            previous = current
            previous_inside = current_inside
    return np.asarray(output, dtype=np.float64)


def _segment_line_intersection(
    start: np.ndarray,
    end: np.ndarray,
    line_start: np.ndarray,
    line_end: np.ndarray,
) -> np.ndarray:
    segment = end - start
    line = line_end - line_start
    denominator = segment[0] * line[1] - segment[1] * line[0]
    if abs(denominator) < 1e-9:
        return end
    offset = line_start - start
    fraction = (offset[0] * line[1] - offset[1] * line[0]) / denominator
    return start + segment * fraction


def _half_plane(point: np.ndarray, start: np.ndarray, end: np.ndarray) -> float:
    return float((end[0] - start[0]) * (point[1] - start[1]) - (end[1] - start[1]) * (point[0] - start[0]))


def _signed_polygon_area(points: np.ndarray) -> float:
    if len(points) < 3:
        return 0.0
    return 0.5 * float(
        np.dot(points[:, 0], np.roll(points[:, 1], -1))
        - np.dot(points[:, 1], np.roll(points[:, 0], -1))
    )
